usr/sbin/p11sak/p11sak.c

```diff
@@ -1142,6 +1192,152 @@ static CK_ATTRIBUTE_TYPE col2type(int col)
+                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp.ulValueLen != sizeof(CK_BBOOL)) {
+                            printf(" Error in retrieving Attribute %s: %lu\n",
+                                    confignode_to_bareval(name)->value, temp.ulValueLen);
+                        } else {
+                            if (long_print) {
+                                printf("          %s: %s\n", confignode_to_bareval(name)->value,
+                                        CK_BBOOL2a(*(CK_BBOOL*) temp.pValue));
+                            } else if (strcmp(CK_BBOOL2a(*(CK_BBOOL*) temp.pValue), "CK_TRUE") == 0) {
+                                printf(" 1 ");
+                                return rc;
+                            }
+                        }    
+                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE) {
+                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
+                        return rc;
+                    } 
+                } else if (strcmp((confignode_to_bareval(type)->value), "CK_ULONG") == 0) {
+                    // build template
+                    CK_ULONG a_ulong;
+                    CK_ULONG us = sizeof(CK_ULONG);
+                    CK_ATTRIBUTE temp = {hex, &a_ulong, us};
+
+                    // get attribute value
+                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+                    if (rc == CKR_OK) {
+                        if (temp.ulValueLen != sizeof(CK_ULONG)) {
+                            printf(" Error in retrieving Attribute %s: %lu\n",
+                                    confignode_to_bareval(name)->value, temp.ulValueLen);
+                        } else {
+                            if (long_print) {
+                                printf("          %s: %lu (0x%lX)\n", confignode_to_bareval(name)->value, 
+                                        *(CK_ULONG*) temp.pValue, *(CK_ULONG*) temp.pValue);    
+                            } else {
+                                printf(" 1 ");
+                                return rc;
+                            }
+                        }    
+                    } else if (rc == CKR_ATTRIBUTE_SENSITIVE) {
+                        printf("          %s: SENSITIVE\n", confignode_to_bareval(name)->value);
+                    } else if (rc != CKR_ATTRIBUTE_TYPE_INVALID) {
+                        return rc;
+                    } 
+                } else if (strcmp((confignode_to_bareval(type)->value), "CK_BYTE") == 0) {
+                    // get length 
+                    CK_ATTRIBUTE temp= {hex, NULL, 0};
+                    rc = funcs->C_GetAttributeValue(session, hkey, &temp, 1);
+
+                    if (temp.ulValueLen != CK_UNAVAILABLE_INFORMATION) {
```

Comment:
I guess I would first check for the return codes like CKR_ATTRIBUTE_SENSITIVE and CKR_ATTRIBUTE_TYPE_INVALID here, before checking for CK_UNAVAILABLE_INFORMATION. Only for CKR_OK yo can assume that ulValueLen is set to either CK_UNAVAILABLE_INFORMATION of the attribute value length. I guess you can omit the check for CK_UNAVAILABLE_INFORMATION then also, since CK_UNAVAILABLE_INFORMATION would only appear with non-CKR_OK return codes...

Reply:
I need the case rc == CKR_OK, but ulValueLen == CK_UNAVAILABLE_INFORMATION, which means the attribute will be skipped. The rest is done.